eotsmanager/randgenerator/randgenerator.go

```diff
@@ -13,7 +13,41 @@ import (
 // GenerateRandomness generates a random scalar with the given key and src
 // the result is deterministic with each given input
 func GenerateRandomness(key []byte, chainID []byte, height uint64) (*eots.PrivateRand, *eots.PublicRand) {
-	// calculate the randomn hash of the key concatenated with chainID and height
+	var randScalar btcec.ModNScalar
+
+	// add iteration as part of HMAC generation to get a uniformly random
+	// value for randScalar via rejection sampling
+	iteration := uint64(0)
+	for {
+		// calculate the random hash with iteration count
+		digest := hmac.New(sha256.New, key)
+		digest.Write(append(append(sdk.Uint64ToBigEndian(height), chainID...), sdk.Uint64ToBigEndian(iteration)...))
+		randPre := digest.Sum(nil)
+
+		// increase iteration count and sample again until overflow or zero does not
+		// happen. It is fine as the chance of overflow or zero is very small (2^-128)
+		overflow := randScalar.SetByteSlice(randPre)
+		if !overflow && !randScalar.IsZero() {
+			break
+		}
+		iteration++
+	}
+
+	// convert the valid scalar into private random
+	privRand := secp256k1.NewPrivateKey(&randScalar)
+	var j secp256k1.JacobianPoint
+	privRand.PubKey().AsJacobian(&j)
+
+	return &privRand.Key, &j.X
+}
+
+// GenerateRandomnessUnsafe generates a random scalar with the given key and src
+// the result is deterministic with each given input
+// NOTE: keeping it is for backward compatibility
+//
+//	this will be deprecated soon
+func GenerateRandomnessUnsafe(key []byte, chainID []byte, height uint64) (*eots.PrivateRand, *eots.PublicRand) {
```

Comment:
How about naming this `GenerateRandomnessLegacy` rather than unsafe? as

- There is only a small bias in the legacy rand generator, and we don't have an exploit for it. So it's a bit too harsh to call it "unsafe"
- It might be a bit confusing with "unsafe" commands that potentially lead to slashing. Using this rand generator won't lead to slashing

Reply:
Make sense!